1-Draft/RFCNNNN-#Requires-Additions.md

```diff
@@ -0,0 +1,155 @@
+---
+RFC: RFC
+Author: Robert Holt
+Status: Draft
+SupercededBy:
+Version:
+Area: Hash-requires
+Comments Due: 2018-05-15
+Plan to implement: Yes
+---
+
+# \#Requires Additions
+
+Currently, PowerShell's `#requires` statement (or perhaps
+pragma?) supports the following parameters:
+
+* `-Version <N>[.<n>]`, where a minimum PowerShell version can be specified
+* `-PSSnapin <PSSnapin-Name> [-Version <N><n>]`, where a required PowerShell Snapin can be specified
+* `-Modules { <Module-name> | <Hashtable> }`, where PowerShell modules that are required can be specified
+* `-ShellId <ShellId>`, where the required Shell ID can be specified
+* `-RunAsAdministrator`, where the script is required to be run as administrator
+
+These features are documented in [about_Requires](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_requires?view=powershell-6), along with the statement that:
+
+> You can use #Requires statements in any script. You
+> cannot use them in functions, cmdlets, or snap-ins.
+
+Currently however, this is untrue, as `#requires`
+statements are allowed by the parser/tokenizer anywhere in
+a script and then effectively hoisted to the top of that
+script to be checked before any part of the script is
+executed, no matter where they are placed.
+
+This RFC proposes the following changes:
+
+* Only allow `#requires` at the top level of a script,
+  before any lines that are not comments (i.e. with the
+  intention that a hashbang can still work, just before
+  any executable PowerShell code). Placing `#requires` anywhere
+  after will cause a parse-time error. This would be a **breaking
+  change**, albeit one that the documentation already claims to be 
+  in force.
```

Comment:
I don't have a strong opinion on introducing a breaking change, but I suppose some might see a benefit in putting `#requires` elsewhere, e.g. at the bottom of the file, perhaps to somewhat hide what might seem like an unimportant implementation detail.

Instead of a breaking change, the documentation could be made clearer. The directive is not scoped - it applies to the outermost ScriptBlock in which the directive appears.

Reply:
Not normative, just some history: Originally #requires had to be the first non-space non-comment line in the file. This is because we were asked to consider the case were we got the language sufficiently wrong that we had to introduce incompatible changes. In that scenario, #requires -version X might cause a different parser to be used for the file. (This feature, along with numbers in the extensions (.ps1), and snapins were  all part of the pre v1 versioning reviews.)  #requires was also done in a preprocessing phase, rather than as part of the main parse. I think @lzybkr changed this in the V3 parser rewrite.